File: android/lib/core/sync_manager.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'directory_watcher.dart';
import 'protocol_frame.dart';
import 'session_manager.dart';
import 'sync_indexer.dart';
import 'sync_model.dart';
import 'transfer_manager.dart';

class SyncManager {
  final String syncFolderPath;
  final TransferManager transferManager;
  final DirectoryWatcher? watcher;

  final StreamController<SyncDifference> _diffController =
      StreamController<SyncDifference>.broadcast();
  final StreamController<String> _syncCompletedController =
      StreamController<String>.broadcast();

  Stream<SyncDifference> get diffStream => _diffController.stream;
  Stream<String> get syncCompletedStream => _syncCompletedController.stream;

  StreamSubscription<String>? _watcherSub;

  SyncManager({
    required this.syncFolderPath,
    required this.transferManager,
    this.watcher,
  }) {
    final dir = Directory(syncFolderPath);
    if (!dir.existsSync()) {
      dir.createSync(recursive: true);
    }

    if (watcher != null) {
      _watcherSub = watcher!.changeStream.listen(_onDirectoryChanged);
      watcher!.start();
    }
  }

  void _onDirectoryChanged(String path) {
    // Directory change detected
  }

  Future<void> requestManifest(
    SessionConnection conn, [
    String folderId = 'default',
  ]) async {
    final req = SyncManifestMessage(
      folderId: folderId,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      entries: [],
    );
    conn.sendJson(MessageType.syncManifestRequest, req.toJson());
  }

  Future<void> sendManifest(
    SessionConnection conn, [
    String folderId = 'default',
  ]) async {
    final manifest = await SyncIndexer.generateManifest(
      syncFolderPath,
      folderId,
    );
    conn.sendJson(MessageType.syncManifestResponse, manifest.toJson());
  }

  Future<SyncDifference?> handleIncomingFrame(
    SessionConnection conn,
    ProtocolFrame frame,
  ) async {
    switch (frame.type) {
      case MessageType.syncManifestRequest:
        try {
          final json =
              jsonDecode(utf8.decode(frame.payload)) as Map<String, dynamic>;
          final req = SyncManifestMessage.fromJson(json);
          await sendManifest(
            conn,
            req.folderId.isEmpty ? 'default' : req.folderId,
          );
        } catch (_) {}
        return null;

      case MessageType.syncManifestResponse:
        try {
          final json =
              jsonDecode(utf8.decode(frame.payload)) as Map<String, dynamic>;
          final remoteManifest = SyncManifestMessage.fromJson(json);

          final localManifest = await SyncIndexer.generateManifest(
            syncFolderPath,
            remoteManifest.folderId,
          );
          final diff = SyncIndexer.calculateDifference(
            localManifest,
            remoteManifest,
          );

          _diffController.add(diff);
          _syncCompletedController.add(remoteManifest.folderId);
          return diff;
        } catch (_) {
          return null;
        }

      case MessageType.syncPullRequest:
        try {
          final json =
              jsonDecode(utf8.decode(frame.payload)) as Map<String, dynamic>;
          final pullReq = SyncPullRequestMessage.fromJson(json);

          for (final rel in pullReq.relativePaths) {
            final localPath = '$syncFolderPath/${rel.replaceAll('\\', '/')}';
            final file = File(localPath);
            if (file.existsSync()) {
              await transferManager.sendFile(conn, localPath);
            }
          }
        } catch (_) {}
        return null;

      default:
        return null;
    }
  }

  void dispose() {
    _watcherSub?.cancel();
    _diffController.close();
    _syncCompletedController.close();
    watcher?.dispose();
  }
}

File: windows/Hinge.Core/SyncManager.cs
using System.Text.Json;

namespace Hinge.Core;

public class SyncManager : IDisposable
{
    private readonly string _syncFolderPath;
    private readonly TransferManager _transferManager;
    private readonly IDirectoryWatcher? _watcher;
    private bool _disposed;

    public string SyncFolderPath => _syncFolderPath;
    public event EventHandler<SyncDifference>? SyncDifferenceComputed;
    public event EventHandler<string>? SyncCompleted;

    public SyncManager(
        string syncFolderPath,
        TransferManager transferManager,
        IDirectoryWatcher? watcher = null)
    {
        _syncFolderPath = syncFolderPath ?? throw new ArgumentNullException(nameof(syncFolderPath));
        _transferManager = transferManager ?? throw new ArgumentNullException(nameof(transferManager));
        _watcher = watcher;

        if (!Directory.Exists(_syncFolderPath))
        {
            Directory.CreateDirectory(_syncFolderPath);
        }

        if (_watcher != null)
        {
            _watcher.Changed += OnDirectoryChanged;
            _watcher.Start();
        }
    }

    private void OnDirectoryChanged(object? sender, string path)
    {
        // Directory change captured
    }

    public async Task RequestManifestAsync(SessionConnection conn, string folderId = "default")
    {
        var req = new SyncManifestMessage { FolderId = folderId };
        await conn.SendJsonAsync(MessageType.SyncManifestRequest, req);
    }

    public async Task SendManifestAsync(SessionConnection conn, string folderId = "default")
    {
        var manifest = await SyncIndexer.GenerateManifestAsync(_syncFolderPath, folderId);
        await conn.SendJsonAsync(MessageType.SyncManifestResponse, manifest);
    }

    public async Task<SyncDifference?> HandleIncomingFrameAsync(SessionConnection conn, ProtocolFrame frame)
    {
        switch (frame.Type)
        {
            case MessageType.SyncManifestRequest:
            {
                var req = JsonSerializer.Deserialize<SyncManifestMessage>(frame.Payload);
                await SendManifestAsync(conn, req?.FolderId ?? "default");
                return null;
            }
            case MessageType.SyncManifestResponse:
            {
                var remoteManifest = JsonSerializer.Deserialize<SyncManifestMessage>(frame.Payload);
                if (remoteManifest == null) return null;

                var localManifest = await SyncIndexer.GenerateManifestAsync(_syncFolderPath, remoteManifest.FolderId);
                var diff = SyncIndexer.CalculateDifference(localManifest, remoteManifest);

                SyncDifferenceComputed?.Invoke(this, diff);
                SyncCompleted?.Invoke(this, remoteManifest.FolderId);
                return diff;
            }
            case MessageType.SyncPullRequest:
            {
                var pullReq = JsonSerializer.Deserialize<SyncPullRequestMessage>(frame.Payload);
                if (pullReq != null)
                {
                    foreach (var rel in pullReq.RelativePaths)
                    {
                        string localFile = Path.Combine(_syncFolderPath, rel.Replace('/', Path.DirectorySeparatorChar));
                        if (File.Exists(localFile))
                        {
                            await _transferManager.SendFileAsync(conn, localFile);
                        }
                    }
                }
                return null;
            }
            default:
                return null;
        }
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        if (_watcher != null)
        {
            _watcher.Changed -= OnDirectoryChanged;
            _watcher.Dispose();
        }
    }
}
